smartsim/_core/utils/serialize.py

```diff
@@ -0,0 +1,237 @@
+    task_id: t.Optional[str],
+    managed: t.Optional[bool],
+    out_file: str,
+    err_file: str,
+    telemetry_data_path: Path,
+) -> t.Dict[str, t.Any]:
+    colo_settings = (model.run_settings.colocated_db_settings or {}).copy()
+    db_scripts = t.cast("t.List[DBScript]", colo_settings.pop("db_scripts", []))
+    db_models = t.cast("t.List[DBModel]", colo_settings.pop("db_models", []))
+    return {
+        "name": model.name,
+        "path": model.path,
+        "exe_args": model.run_settings.exe_args,
+        "run_settings": _dictify_run_settings(model.run_settings),
+        "batch_settings": _dictify_batch_settings(model.batch_settings)
+        if model.batch_settings
+        else {},
+        "params": model.params,
+        "files": {
+            "Symlink": model.files.link,
+            "Configure": model.files.tagged,
+            "Copy": model.files.copy,
+        }
+        if model.files
+        else {
+            "Symlink": [],
+            "Configure": [],
+            "Copy": [],
+        },
+        "colocated_db": {
+            "settings": colo_settings,
+            "scripts": [
+                {
+                    script.name: {
+                        "backend": "TORCH",
+                        "device": script.device,
+                    }
+                }
+                for script in db_scripts
+            ],
+            "models": [
+                {
+                    model.name: {
+                        "backend": model.backend,
+                        "device": model.device,
+                    }
+                }
+                for model in db_models
+            ],
+        }
+        if colo_settings
+        else {},
+        "telemetry_metadata": {
+            "status_dir": str(telemetry_data_path),
+            "step_id": step_id,
+            "task_id": task_id,
+            "managed": managed,
+        },
+        "out_file": out_file,
+        "err_file": err_file,
+    }
+
+
+def _dictify_ensemble(
+    ens: Ensemble,
+    members: t.Sequence[t.Tuple[Model, TStepLaunchMetaData]],
+) -> t.Dict[str, t.Any]:
+    return {
+        "name": ens.name,
+        "params": ens.params,
+        "batch_settings": _dictify_batch_settings(ens.batch_settings)
+        # FIXME: Typehint here is wrong, ``ens.batch_settings`` can
+        # also be an empty dict for no discernible reason...
+        if ens.batch_settings else {},
+        "models": [
+            _dictify_model(model, *launching_metadata)
+            for model, launching_metadata in members
+        ],
+    }
+
+
+def _dictify_run_settings(run_settings: RunSettings) -> t.Dict[str, t.Any]:
+    return {
+        "exe": run_settings.exe,
+        # TODO: We should try to move this back
+        # "exe_args": run_settings.exe_args,
+        "run_command": run_settings.run_command,
+        "run_args": run_settings.run_args,
+        # TODO: We currently do not have a way to represent MPMD commands!
```

Comment:
This doesn't seem like it should be that hard to do? MPMD should just be lists of the extra slurm/pbs commands, e.g. `-N 1 -n 1 -c 32 : -N 1 -n32 -c 1`

Reply:
I would agree if the dashboard were expecting a `str`. Unfortunately right now the dashboard expects a `dict[str, str]` and we cannot "re-add" keys for the dashboard to display (for e.g. in your example it would be non-trivial to display the two different values of `-n`).

We could turn the `"run_args"` key into a `list[dict[str, str]]`, but I would have to check with @AlyssaCote to see how much effort it would take for the dashboard to render this info on the frontend!

Reply:
Ahhhh, I see. Ok, as a stop gap for now let's throw a warning here and put up a ticket.